app/coffee/Features/Labels/LabelsHandler.coffee

```diff
@@ -0,0 +1,56 @@
+ProjectEntityHandler = require "../Project/ProjectEntityHandler"
+DocumentUpdaterHandler = require('../DocumentUpdater/DocumentUpdaterHandler')
+Async = require('async')
+
+
+module.exports = LabelsHandler =
+
+	labelCaptureRegex: () ->
+		/\\label\{([^\}\n\\]{0,80})\}/g
+
+	getAllLabelsForProject: (projectId, callback=(err, projectLabels)->) ->
+		ProjectEntityHandler.getAllDocs projectId, (err, docs) ->
+			if err?
+				return callback(err)
+			LabelsHandler.extractLabelsFromProjectDocs docs, (err, projectLabels) ->
+				if err?
+					return callback(err)
+				callback(null, projectLabels)
+
+	getLabelsForDoc: (projectId, docId, callback=(err, docLabels)->) ->
+		# Flush doc first, because this action is often performed while
+		# a document is being edited by the client. By contrast,
+		# `getAllLabelsForProject` is called only when a project/editor is loaded
```

Comment:
I think we should actually flush in both cases, i.e. also in `getAllLabelsForProject`